### PR TITLE
Bug 1881604 - Helper for Transforming BCP 47 code to defined Language

### PR DESCRIPTION
**Pull request by ohall-m**
This patch adds a helper for converting BCP 47 codes to translations supported languages.

Adds:
* `toLanguageMap` - maps a map of BCP 47 code (key) to Language (value)
* `findLanguage` - finds a language using the `toLanguageMap`
* `mapLanguageSettings` - makes a new display ready map for converting the language settings from Map<String, LanguageSetting>?) to Map<Language?, LanguageSetting>

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1881604

### CODE REVIEW

**Comment by ohall-m** [comment]
Expected use case is for getting `pageSettings` in display ready format on this [screen](https://www.figma.com/file/GT6oZLFw32LiCwKl5GOMWD/Full-Page-Translation?type=design&node-id=7202-113172&mode=design&t=7A2uDDSwa4kPkZ1p-0).

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationSupport.kt [hunk 1]
```diff
@@ -16,3 +16,53 @@ data class TranslationSupport(
...
+/**
+ * Convenience method to convert a language setting map using a BCP 47 code as a key to a map using
+ * [Language] as a key.
+ *
+ * @param languageSettings The map of language settings, where the key [String] is a BCP 47 code.
+ */
+fun TranslationSupport.mapLanguageSettings(
+    languageSettings: Map<String, LanguageSetting>?,
+): Map<Language?, LanguageSetting>? {
+    return languageSettings?.mapKeys { findLanguage(it.key) }?.filter { it.key != null }
```
Comment (gabrielluong) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationSupport.kt:67]:
```suggestion
    return languageSettings?.mapKeys { findLanguage(it.key) }?.filterKeys { it != null }
```

**Review by gabrielluong — APPROVED** [review]

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationSupport.kt [hunk 2]
```diff
@@ -16,3 +16,53 @@ data class TranslationSupport(
...
+
+    if (toLanguagesMap == null && fromLanguagesMap != null) {
+        return fromLanguagesMap
+    }
+
+    if (toLanguagesMap != null && fromLanguagesMap != null) {
+        return toLanguagesMap + fromLanguagesMap
+    }
+
+    return null
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationSupport.kt]: Thanks, much neater! Compiler was able to offer one step further too.

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationSupport.kt [hunk 3]
```diff
@@ -16,3 +16,53 @@ data class TranslationSupport(
...
+}
+
+/**
+ * Convenience method to find a [Language] given a BCP 47 language code.
+ *
+ * @param languageCode The BCP 47 language code.
+ *
+ * @return The [Language] associated with the language code or null.
+ */
+fun TranslationSupport.findLanguage(languageCode: String?): Language? {
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationSupport.kt]: Good spot! Was able to fix `languageCode` to non-optional. `Language?`is coming from the possibility of an empty map.

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationSupport.kt (hunk 1 above)
Comment (ohall-m) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationSupport.kt]: Nice, ty!

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationSupport.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (gabrielluong) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationSupport.kt]: Oh yes, that's actually what i intended.

**Pull request merged**